libexec/rtld-elf/rtld_c18n.c

```diff
@@ -745,14 +745,23 @@ init_stk_table(struct stk_table *table, struct tcb_wrapper *wrap)
 	};
 
 	/*
-	 * Record RTLD's stack in the stack lookup table. The recorded size of
-	 * the stack remains zero so that the stack is not ummapped upon thread
-	 * exit.
+	 * Record RTLD's stack in the stack lookup table.
 	 */
 #ifdef __ARM_MORELLO_PURECAP_BENCHMARK_ABI
 	sp = cheri_setoffset(cheri_getstack(), 0);
 #else
-	sp = NULL;
+	/*
+	 * RTLD's actual stack is the Executive stack which does not need to be
+	 * stored in the stack lookup table. Instead, fill the table entry with
+	 * a one-byte dummy stack.
+	 *
+	 * Note that NULL cannot be used here because a compartment can then
+	 * pretend to be RTLD when a signal is delivered. Nor can a zero-length
+	 * stack be used because the signal handler uses the recorded size of
+	 * the stack to determine whether it has been allocated.
+	 */
+	static char dummy_stk;
```

Comment:
This is ugly, surely there's a better way. Where does this end up being used to determine that the compartment is RTLD, anyway? Isn't the trusted stack tracking the compartment ID?

Reply:
> Isn't the trusted stack tracking the compartment ID?

This is generally true except at certain points in the trampoline where the trusted stack becomes temporarily out-of-sync with the currently installed untrusted stack.

The signal handler needs to handle such situations and makes a few guesses about the actual compartment that the current untrusted stack belongs to. It does so by checking whether the current untrusted stack is a subset of one of the candidate compartments' stack. Hence the dummy stack must be tagged and have positive length to prevent compartments from impersonating as RTLD.

Reply:
"makes a few guesses about the actual compartment that the current untrusted stack belongs to" doesn't fill me with confidence as to the security of this mechanism

Reply:
The security argument definitely has lots of room for improvement. But at least signal handling should be functionally correct now.